doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,230 @@
+4. STORAGES receives ACK OK response from the Coordinator.
+
+This is fine and STORAGE can just ignore the ACK OK response.
+
+## Alternatives
+
+- We don't hold read capabilities at the global timestamp. Instead, we block all queries that require a timestamp larger
+  than the global timestamp until the global timestamp advances to the timestamp of that
+  query. [#11150](https://github.com/MaterializeInc/materialize/pull/11150)
+  and [#10173](https://github.com/MaterializeInc/materialize/pull/10173) are previous pull requests that try and
+  implement this query blocking.
+- If we didn't hold read capabilities, then we could always advance the global timestamp to the timestamp of most recent
+  operation. Any operation on objects that haven't caught up to the global timestamp must be blocked until the object
+  catches up to the global timestamp.
+- Instead of explicitly holding read capabilities at the global timestamp, the Coordinator can include the current
+  global timestamp in every message sent to dataflow nodes. The nodes would then prevent compaction up to the most
+  recently seen global timestamp. After the node learns of a new timestamp they can allow compaction of every object up
+  to the new timestamp. However, if a node doesn't receive any messages from the Coordinator then memory usage will
+  balloon up until it hears from the Coordinator again.
+- Instead of using group commit/write coalesce, the coordinator could buffer all writes to user tables in a durable
+  write ahead log. All writes to user tables that occur in the same millisecond can be assigned the same timestamp, as
+  long as there are no reads in-between them. Then at the end of the millisecond, or after X milliseconds, all writes
+  are sent to STORAGE in a batch. If Materialize crashes/restarts before sending the writes to STORAGE, then it can just
+  read the unsent writes from the durable logs after starting back up. This prevents us from having to increment the
+  global timestamp on every write to a user table. However, this does not solve the problem if there are reads
+  in-between the writes or if we read from a view that's ahead of the wall clock.
+- Hybrid Logical Timestamps (HLT) are timestamps where the X most significant bits are used for the epoch and the Y
+  least significant bits are used for a counter. Every clock tick the X bits are increased by 1 and the Y bytes are all
+  reset to 0. Every write increments the Y bits by 1. This would allow us to have 2^Y writes per clock tick per object,
+  without blocking any queries. Any write in excess of 2^Y per clock tick per object must be blocked until the next
+  clock tick. This way no object's `upper` would have to advance past the current global timestamp if the global
+  timestamp tracked the wall clock. CockroachDB uses similar timestamps
+  here: https://github.com/cockroachdb/cockroach/blob/711ccb5b8fba4e6a826ed6ba46c0fc346233a0f7/pkg/sql/sem/builtins/builtins.go#L8548-L8560
```

Comment:
are we going to explore this more?

Reply:
I think it would be interesting to explore more in general but I'm probably not going to do this as part of this design doc. There was more detail in an older commit here: https://github.com/MaterializeInc/materialize/blob/d87df3b729de31a17bbdbd63c06277155ae496e8/doc/developer/design/20220516_transactional_consistency.md?plain=1#L374-L406

Every node is responsible for generating their own timestamps and we don't provide any guarantees about the accuracy of each clock. This makes it extremely difficult/impossible to compare timestamps across nodes, which is something that would be extremely useful when dealing with linearizability. We mostly get around this by serializing everything through the Coordinator. However, I used to have an additional bullet point here about changing the type of clock to something that would allow us to accurately compare timestamps. Those were
- Hybrid Logical Clocks (which this bullet point alludes to)
- Vector Logical Clocks
- Timestamp Oracle (based on the Percolator design)
I thought it'd be interesting to think about those, but I think ultimately it would be a much bigger change than the design from this doc.